Sources/protoc-gen-grpc-swift/Generator-Client.swift

```diff
@@ -22,35 +22,106 @@ extension Generator {
     println()
     printServiceClientProtocol()
     println()
+    printClientProtocolExtension()
+    println()
     printServiceClientImplementation()
   }
 
-  private func printServiceClientProtocol() {
-    println("/// Usage: instantiate \(clientClassName), then call methods of this protocol to make API calls.")
-    println("\(options.visibility.sourceSnippet) protocol \(clientProtocolName) {")
-    indent()
-    for method in service.methods {
-      self.method = method
-      switch streamingType(method) {
-      case .unary:
-        println("func \(methodFunctionName)(_ request: \(methodInputName), callOptions: CallOptions?) -> UnaryCall<\(methodInputName), \(methodOutputName)>")
+  private func printFunction(name: String, arguments: [String], returnType: String, access: String? = nil, bodyBuilder: (() -> ())?) {
+    // Add a space after access, if it exists.
+    let accessOrEmpty = access.map { $0 + " " } ?? ""
 
-      case .serverStreaming:
-        println("func \(methodFunctionName)(_ request: \(methodInputName), callOptions: CallOptions?, handler: @escaping (\(methodOutputName)) -> Void) -> ServerStreamingCall<\(methodInputName), \(methodOutputName)>")
+    let hasBody = bodyBuilder != nil
 
-      case .clientStreaming:
-        println("func \(methodFunctionName)(callOptions: CallOptions?) -> ClientStreamingCall<\(methodInputName), \(methodOutputName)>")
+    if arguments.isEmpty {
+      // Don't bother splitting across multiple lines if there are no arguments.
+      self.println("\(accessOrEmpty)func \(name)() -> \(returnType)", newline: !hasBody)
+    } else {
+      self.println("\(accessOrEmpty)func \(name)(")
+      self.withIndentation {
+        // Add a comma after each argument except the last.
+        arguments.forEach(beforeLast: {
+          self.println($0 + ",")
+        }, onLast: {
+          self.println($0)
+        })
+      }
+      self.println(") -> \(returnType)", newline: !hasBody)
+    }
 
-      case .bidirectionalStreaming:
-        println("func \(methodFunctionName)(callOptions: CallOptions?, handler: @escaping (\(methodOutputName)) -> Void) -> BidirectionalStreamingCall<\(methodInputName), \(methodOutputName)>")
+    if let bodyBuilder = bodyBuilder {
+      self.println(" {")
+      self.withIndentation {
+        bodyBuilder()
+      }
+      self.println("}")
+    }
+  }
+
+  private func printServiceClientProtocol() {
+    self.println("/// Usage: instantiate \(self.clientClassName), then call methods of this protocol to make API calls.")
+    self.println("\(self.access) protocol \(self.clientProtocolName): GRPCClient {")
```

Comment:
Do we want to continue putting the access level of this protocol in the generated code? Or would it be better just to put it on all the methods? This gets us a bit closer to NIO, but I don't know if we want to do that.

Reply:
It would also avoid the need for `accessOrEmpty` in `printFunction`, as all other call sites have the access passed.

Reply:
For protocol declaration it _has_ to be on the `protocol` rather than the method sadly :(